### ISSUE TITLE
[Feature Request] Stronger adaptability mode than "Aggressive"

### ISSUE THREAD

**Issue by PinkSerenity** [EmrecanKaracayir/Codemos-Modern#208]
Hii,

I find the "Aggressive" mode still too weak for my tastes. Is there an easy way to do even stronger theming? If not, can "Hyper-aggressive" please be a mode?

Thank you very much.

**Comment by EmrecanKaracayir** [comment]
Version 3 is pretty much ready. And it will ship with that feature and many more, stay tuned!

**Comment by PinkSerenity** [comment]
> To achieve it, download preview and start configuring. When asked about adaptation intensity, select custom and enter a number between 0 to 100 (percent).

I just tried it, but 100% is still too weak for my tastes. Keep in mind that my previous theme looked like this:
![image](https://github.com/user-attachments/assets/f6b74a09-470b-40eb-8d25-cf7fc0e7de32)

In Windows Terminal I am using a color with an HSV value of (320°, 16%, 99%), which has twice the saturation of the old Aggressive. So if you'd allow us to go over the 100%, that would be great.

**Comment by EmrecanKaracayir** [comment]
can you share your configuration (copy all codemos modern configurations under user settings) So I can better understand what is happening?

**Comment by EmrecanKaracayir** [comment]
![Adaptation](https://github.com/user-attachments/assets/907e5226-b5fb-47de-a8f8-2e4349984764)

**Comment by PinkSerenity** [comment]
> Can you share your configuration (codemosModern.* configurations under user settings) So I can better understand what is happening?

Sure:
```json
{
    "codemosModern.light.design": "flat",
    "codemosModern.light.accentColor": "#AC0074",
    "codemosModern.light.adaptationColor": "#AC0074",
    "codemosModern.light.adaptationIntensity": 100,

    "codemosModern.dark.design": "flat",
    "codemosModern.dark.accentColor": "#ED63BA",
    "codemosModern.dark.adaptationColor": "#ED63BA",
    "codemosModern.dark.adaptationIntensity": 100,

    "window.autoDetectColorScheme": true,
}
```

> This new adaptation algorithm is scientifically blending two colors (backdrop and source), based on the source alpha (opacity) value, which can only be in between 0 to 1, this is why I ask for a percentage.
>
> Because of the nature of colors blending, bright colors (sources) on bright surfaces (backdrops) blend better and dark colors on dark surfaces blend better. In your example, dimmer color (your backdrop) trying to blend with a brighter color (source or UI) resulting with low saturation.

I see. I think. How exactly does it blend them? If the adaption color is the backdrop, why does a higher source opacity equal stronger adaption? That's the part I unfortunately don't understand yet. If the source gets drawn over the backdrop, I'd assume (with my limited knowledge) that 0% is full backdrop and 100% is full source. But the algorithm seems to be more complex then that.

If you don't have time or leisure explaining that, I completely understand. Please don't feel pressured.

**Comment by EmrecanKaracayir** [comment]
> How exactly does it blend them?

I am using Luminosity blend mode. Here comes the not so scientific explanation, In essence, the source color's brightness adjusts the backdrop color's lightness, while the backdrop retains its original hue and saturation.

When you use a light source color, like light gray, on a light backdrop color, such as yellow, the resulting blend maintains the yellow's hue and saturation because their brightness levels are similar. This means the yellow retains its vividness and looks much like the original color.

However, when you use the same light source color on a darker backdrop, like blue, the blend results in a color with the same blue hue but significantly altered brightness. This change can make the blue appear less saturated and vibrant because the contrast between the light source and dark backdrop affects the overall luminosity. Essentially, matching brightness levels helps preserve color saturation, while mismatched levels can lead to more drastic changes in appearance.

#### Example
![Yellow vs Blue](https://github.com/user-attachments/assets/0557f5b8-2ebf-4945-8631-dc6816e5767a)
As you can see yellow pretty much stays at yellow but blue changes drastically.

> If the adaption color is the backdrop, why does a higher source opacity equal stronger adaption?

It is reversed when calculating, so 100 means 0 percent of the source.

> But the algorithm seems to be more complex then that

The previous screenshot I posted which was the yellow theme actually showcases the algorithm, not so complex. In fact, it was a lot simpler than the previous algorithm.

**Comment by EmrecanKaracayir** [comment]
This is the result of changing `palette.base` in `light/flat.ts` from `#F9F9F9` to `#CCCCCC`

![Screenshot 2024-12-10 at 19 24 15](https://github.com/user-attachments/assets/4141cd4d-af50-4473-b9cb-2a7de1b9addb)

We can maybe make base color customizable but I am not able to predict the outcome on other stuff.

**Comment by PinkSerenity** [comment]
Hii. Sorry that I haven't responsed yesterday, I had a lot on my mind. But I kinda kept experimenting and got good results with applying alpha on the backdrop and using `mix-blend-mode: soft-light;`. This is how it would look like:

![image](https://github.com/user-attachments/assets/83e0b6e2-e4a4-4353-bb69-2568c9f27600)
The order is 5%, 10%, 20%, 35% and 50%.

Here the same with the default Yellow (both light and dark values):
![image](https://github.com/user-attachments/assets/6a7945fa-9d66-4dfe-8561-71f171bb2836)


I will keep experimenting though.

**Comment by EmrecanKaracayir** [comment]
Interesting findings. The reason I chose Luminosity blend is that I had to control the brightness of the overall UI elements for accessibility reasons. For the light theme, In worst case scenario (backdrop is black) and white black text must be legible on the resulting mimic layer (adapted color). Please experiment with worst case scenarios and see how soft-light blend mode handles these cases.

Also if you are interested, this is the [official wcag documents](https://www.w3.org/TR/compositing-1/#blending) on the algorithms used to achieve blend modes.

**Comment by PinkSerenity** [comment]
For the record, here are some more exampleswith lower numbers:
![image](https://github.com/user-attachments/assets/c4866dd2-adea-49e0-9656-0154a796649a)
![image](https://github.com/user-attachments/assets/a44d0812-8cfd-417e-8a37-7e6ab938d785)

**Comment by PinkSerenity** [comment]
![image](https://github.com/user-attachments/assets/f8b0ee1d-409d-4752-b5f7-5c8c6d9e5dd4)
I'd say until 50% it's at least readable, even if it's not pleasant. You display it multiplied by 10, so that 10% becomes 100%, say that everything above 100% is unsupported and entering values above 500% is impossible (but still possible in the `settings.json` for the people who are truly crazy).

**Comment by EmrecanKaracayir** [comment]
Okay, I did a half-baked implementation in the [new branch](https://github.com/Codemos-Inc/Codemos-Modern/tree/exp_soft-light-blend). I changed `intensity` inversely mix source and resulting blend. You can browse the changes in the commit 6f175d12efab5ea71859183524b2b10c872a808d.


#### Here is the vsix package of the experimental version
[codemos-modern-3.0.0rc1.exp.vsix.zip](https://github.com/user-attachments/files/18098658/codemos-modern-3.0.0rc1.exp.vsix.zip)

It needs a lot of testing on my side to decide if it can replace the current method. Also, accessibility is my number one concern.

**Comment by PinkSerenity** [comment]
> It needs a lot of testing on my side to decide if it can replace the current method. Also, accessibility is my number one concern.

Of course. I'm still experimenting and trying to keep some of the brightness control by implementing mutliple backdrops somehow. I have no idea how it will turn out, but I'm hooked.
But thank you really much for actually considering this.

**Comment by PinkSerenity** [comment]
> #### Here is the vsix package of the experimental version
> [codemos-modern-3.0.0rc1.exp.vsix.zip](https://github.com/user-attachments/files/18098658/codemos-modern-3.0.0rc1.exp.vsix.zip)

I really like it, but I would hide the stronger settings from normal access. Accessibility is good, and people shouldn't stumble upon a way to make their editor unreadable.

I will keeping looking for an alternative way that keeps brightness at a readable level at all values.

**Comment by PinkSerenity** [comment]
I can't believe I was making it so complicated all the time.
```css
.backdrop.bright {
	background-color: hsl(from #AC0074 h s max(20, l) / var(--alpha));
}

.backdrop.dark {
	background-color: hsl(from #ED63BA h s min(80, l) / var(--alpha));
}
```
It looks like this:
![image](https://github.com/user-attachments/assets/b366473e-592b-4237-bf54-bef9b9198922)
Adaption color: Pink (Light / Dark variant)

![image](https://github.com/user-attachments/assets/b4d3286a-9fbf-4428-a6ee-8b4fa57bf000)
Adaption color: Yellow (Light / Dark variant)

![image](https://github.com/user-attachments/assets/312ad107-4c8f-4022-84fe-a3d68caf5e33)
Adaption color: Black, White


You can set the `max` and `min` to any value you like. I think 20 and 80 is still a bit low, the text is barely readable in the 100% one. Maybe 30 and 70?
![image](https://github.com/user-attachments/assets/32959f84-8776-4fb2-906a-5a4036cefbb9)
